src/registers/debug.rs

```diff
@@ -0,0 +1,580 @@
+//! Functions to read and write debug registers.
+
+#![deny(unsafe_op_in_unsafe_fn)]
+
+use core::convert::TryFrom;
+use core::ops::Range;
+
+use crate::VirtAddr;
+
+use bit_field::BitField;
+use bitflags::bitflags;
+
+/// Debug Address Register
+///
+/// Holds the address of a hardware breakpoint.
+pub trait DebugAddressRegister {
+    /// Reads the current breakpoint address.
+    fn read() -> VirtAddr;
+
+    /// Writes the provided breakpoint address.
+    fn write(addr: VirtAddr);
+}
+
+macro_rules! debug_address_register_struct {
```

Comment:
If we're going to bother with using a macro here, it seems like we should use the macro to entirely avoid repetition, not just avoid repetition for the struct declaration. 

Could we do something similar to what we do for the [segment registers](https://github.com/rust-osdev/x86_64/blob/d7e62d2490507b7f75542dd8a8ac749be14e4c47/src/instructions/segmentation.rs#L37-L56)

Ideally, we would just be able to have a macro `debug_address_register`, and write:
```rust
debug_address_register!(Dr0, "dr0");
debug_address_register!(Dr1, "dr1");
debug_address_register!(Dr2, "dr2");
debug_address_register!(Dr3, "dr3");
```

The implementation would look something like:
```rust
macro_rules! debug_address_register {
    ($Dr:ident, $name:literal) => {
        pub struct $Dr;

        #[cfg(feature = "instructions")]
        impl DebugAddressRegister for $Dr {
            #[inline]
            fn read() -> VirtAddr {
                let addr;
                unsafe {
                    asm!(concat!("mov {}, ", $name), out(reg) addr, options(nostack, preserves_flags));
                }
                VirtAddr::new(addr)
            }

            #[inline]
            fn write(addr: VirtAddr) {
                unsafe {
                    asm!(concat!("mov ", $name, ", {}"), in(reg) addr.as_u64(), options(nostack, preserves_flags));
                }
            }
        }
    };
}
```

Reply:
Done :+1: